services/common/rs_server_common/authentication/oauth2.py

```diff
@@ -0,0 +1,314 @@
+# Copyright 2023-2024, CS GROUP - France, https://www.csgroup.eu/
+#
+# This file is part of APIKeyManager project
+#     https://github.com/csgroup-oss/apikey-manager/
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Annotated
+from urllib.parse import parse_qs, urlencode, urlparse, urlunparse
+
+from authlib.integrations import starlette_client
+from authlib.integrations.starlette_client.apps import StarletteOAuth2App
+from fastapi import (
+    APIRouter,
+    Depends,
+    FastAPI,
+    HTTPException,
+    Request,
+    Response,
+    status,
+)
+from fastapi.responses import HTMLResponse
+from rs_server_common import settings
+from rs_server_common.authentication.keycloak_util import KCUtil
+from rs_server_common.utils.utils2 import AuthInfo
+from starlette.config import Config as StarletteConfig
+from starlette.middleware import Middleware
+from starlette.middleware.sessions import SessionMiddleware
+from starlette.requests import Request
+from starlette.responses import RedirectResponse
+
+# authlib object for oauth2 keycloak authentication
+KEYCLOAK: StarletteOAuth2App = None
+
+# Const values
+AUTH_PREFIX = "/auth"
+LOGIN_FROM_BROWSER = "/login"
+LOGIN_FROM_CONSOLE = "/login_from_console"
+COOKIE_NAME = "user"
+SWAGGER_HOMEPAGE = "/docs"  # swagger /docs page as called from the cluster
+
+
+class LoginAndRedirect(Exception):
+    """
+    Used to call the login endpoint and redirect to the calling endpoint.
+    See https://github.com/fastapi/fastapi/discussions/7817#discussioncomment-5144391
+    """
+
+    pass
+
+
+async def is_from_browser(request: Request) -> bool:
+    """Return True if the request comes from a browser, False if from a console (curl, python console, ...)"""
+
+    # See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent
+    try:
+        user_agent = request.headers["user-agent"]
+    except KeyError:
+        return False
+    return any(
+        browser in user_agent
+        for browser in (
```

Comment:
We have to define all possible browsers here? Doesn't look maintainable.

Reply:
I've looked at other solutions to determine if we're trying to log from a browser or console, and this one seems to be the more secure. I don't think we'll need to update this list because all existing browsers are already in the list, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent#browser_compatibility